### PR TITLE
Bump k8s.io/client-go to 1.16

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: galley/pkg/config/source/kube/apiserver/source.go [hunk 1]
```diff
@@ -101,13 +101,13 @@ func (s *Source) Dispatch(h event.Handler) {
 // Start implements processor.Source
 func (s *Source) Start() {
 	s.mu.Lock()
-	defer s.mu.Unlock()
```
Comment (hzxuzhonghu) [inline galley/pkg/config/source/kube/apiserver/source.go]: @howardjohn This is the root cause,  race with `s.onCrdEvent`, which also try to acquire lock.

**Comment by hzxuzhonghu** [comment]
fixes: #17831

File: galley/pkg/config/source/kube/apiserver/watcher.go [hunk 2]
```diff
@@ -85,15 +85,12 @@ func (w *watcher) start() {
...
-			w.handler.Handle(event.FullSyncFor(w.resource.Collection.Name))
-		}
-	}()
-
 	// Start CRD shared informer and wait for it to exit.
 	go informer.Run(done)
+	// Send the FullSync event after the cache syncs.
+	if cache.WaitForCacheSync(done, informer.HasSynced) {
+		go w.handler.Handle(event.FullSyncFor(w.resource.Collection.Name))
+	}
```
Comment (ayj) [inline galley/pkg/config/source/kube/apiserver/watcher.go]: Was this change also necessary to fix the race? It serializes the per-resource watch setup.

Comment (hzxuzhonghu) [reply galley/pkg/config/source/kube/apiserver/watcher.go]: Yes,  it has to wait cache sync before exit.

**Comment by hzxuzhonghu** [comment]
The CI passed eventually, the mcp server sync logic is not very well.

**Comment by howardjohn** [comment]
/test unit-tests_istio

Want to make sure it didn't just pass because we got lucky

**Comment by hzxuzhonghu** [comment]
The error log is 
```
=== RUN   TestProcessor_Publishing
2019-10-23T16:22:02.736257Z	info	runtime	Starting processor...
2019-10-23T16:22:02.736274Z	debug	runtime	Starting process loop
2019-10-23T16:22:02.736292Z	debug	runtime	Incoming source event: [Event](FullSync)
2019-10-23T16:22:02.736304Z	info	runtime	Synchronization is complete, starting distribution.
2019-10-23T16:22:02.741941Z	debug	runtime	Incoming source event: [Event](Added: [VKey](empty:r1 @v1))
2019-10-23T16:22:02.742396Z	error	processing	length of collection is less than 4, does not match expectation. collection: [empty]
2019-10-23T16:22:02.742446Z	debug	runtime	In-memory State has changed:
[State @1]
[0] (empty @1)
  [0] (r1)
[1] (struct @0)


2019-10-23T16:22:02.742456Z	debug	runtime	Processor.process: publish
2019-10-23T16:22:02.744470Z	info	runtime	Stopping processor...
2019-10-23T16:22:02.744529Z	debug	runtime	Processor.process: publish
2019-10-23T16:22:02.744837Z	debug	runtime	Process.process: Exiting worker thread
panic: sync: negative WaitGroup counter

goroutine 22 [running]:
sync.(*WaitGroup).Add(0xc0003b0070, 0xffffffffffffffff)
	/usr/local/go/src/sync/waitgroup.go:74 +0x139
sync.(*WaitGroup).Done(...)
	/usr/local/go/src/sync/waitgroup.go:99
istio.io/istio/galley/pkg/runtime.TestProcessor_Publishing.func1()
	/home/prow/go/src/istio.io/istio/galley/pkg/runtime/processor_test.go:184 +0x34
istio.io/istio/galley/pkg/runtime.(*Processor).Start.func2(0x145f1e0, 0xc0004d0080)
	/home/prow/go/src/istio.io/istio/galley/pkg/runtime/processor.go:190 +0x2b6
istio.io/istio/galley/pkg/util.(*Worker).Start.func1(0xc0003a2340, 0xc0004d20b0)
	/home/prow/go/src/istio.io/istio/galley/pkg/util/worker.go:88 +0x43
created by istio.io/istio/galley/pkg/util.(*Worker).Start
	/home/prow/go/src/istio.io/istio/galley/pkg/util/worker.go:86 +0xdf
FAIL	istio.io/istio/galley/pkg/runtime	2.095s
```

**Comment by howardjohn** [comment]
actually maybe not related to this? I saw that failure today on another pr https://github.com/istio/istio/issues/18258

If this is passing normally then maybe it's fine, let's try one more and see

**Comment by hzxuzhonghu** [comment]
I tried 100 times without flake, but thinks the timer reset is not right, maybe related.

**Comment by howardjohn** [comment]
Thanks for working on this

**Comment by howardjohn** [comment]
Fixes https://github.com/istio/istio/issues/17831

**Pull request merged**